gnovm/pkg/gnolang/realm.go

```diff
@@ -1205,6 +1196,24 @@ func copyValueWithRefs(parent Object, val Value) Value {
 		return bl
 	case RefValue:
 		return cv
+	case *HeapItemValue:
+		// NOTE: While this could be eliminated sometimes with some
+		// intelligence prior to persistence, to unwrap the
+		// HeapItemValue in case where the HeapItemValue only has
+		// refcount of 1,
+		//
+		//  1.  The HeapItemValue is necessary when the .Value is a
+		//    primitive non-object anyways, and
+		//  2. This would mean PointerValue.Base is nil, and we'd need
+		//    additional logic to re-wrap when necessary, and
+		//  3. And with the above point, it's not clear the result
+		//    would be any faster.  But this is something we could
+		//    explore after launch.
+		hiv := &HeapItemValue{
+			ObjectInfo: cv.ObjectInfo.Copy(),
```

Comment:
This is more of a general question -- why must the object info be copied in this function?

Reply:
I think, just because we copy the object info everywhere else in this function.

Reply:
It is used by reference *ObjectInfo, so if we don't copy it, the HeapItemValue would end up having the containing value's object ID, instead of there being two distinct ones, and the system would crash after it tramples the value with a heapitemvalue containing... itself.